lib/jekyll/readers/theme_data_reader.rb

```diff
@@ -0,0 +1,31 @@
+# frozen_string_literal: true
+
+module Jekyll
+  class ThemeDataReader < Jekyll::DataReader
+    attr_reader :site, :content
+
+    def read(dir)
+      return unless site.theme&.data_path
+
+      base = site.in_theme_dir(dir)
+      read_data_to(base, @content)
+      Jekyll::Utils.deep_merge_hashes(@content, @site.data)
+    end
+
+    def read_data_to(dir, data)
+      return unless File.directory?(dir) && !@entry_filter.symlink?(dir)
+
+      entries = Dir.chdir(dir) do
+        Dir["*.{yaml,yml,json,csv,tsv}"]
```

Comment:
The superclass also includes reading subdirectories. If we're not going to read subdirectories here, then let's leave a comment saying this only reads one layer deep and why we chose that.